src/Executables/ParallelInfo/ParallelInfo.cpp

```diff
@@ -1,13 +1,105 @@
 // Distributed under the MIT License.
 // See LICENSE.txt for details.
 
-#include <iostream>
+#include "Executables/ParallelInfo/ParallelInfo.hpp"
 
+#include <cmath>
+
+#include "ErrorHandling/Error.hpp"
 #include "Informer/InfoFromBuild.hpp"
+#include "Parallel/Exit.hpp"
+#include "Parallel/Info.hpp"
+#include "Parallel/Printf.hpp"
+
+/// \cond HIDDEN_SYMBOLS
+ParallelInfo::ParallelInfo(CkArgMsg* msg) {
+  Parallel::printf("Executing '%s' using %d processors.\n", msg->argv[0],
+                   Parallel::number_of_procs());
+  if (msg->argc > 1) {
+    std::stringstream error_msg;
+    error_msg << "Expected zero command line options, not " << msg->argc - 1
+              << "\nReceived:\n";
+    for (int i = 1; i < msg->argc; ++i) {
+      error_msg << "'" << msg->argv[i] << "'\n\n";
+    }
+    ERROR(error_msg.str());
+  }
+
+  Parallel::printf("%s\n", info_from_build());
+
+  Parallel::printf(
+      "\nHeader:\n"
+      "[1]Node ID            (my_node)\n"
+      "[2]PE ID              (my_proc)\n"
+      "[3]Number Of Procs    (number_of_procs)\n"
+      "[4]Number Of Nodes    (number_of_nodes)\n"
+      "[5]Procs On Node      (procs_on_node)\n"
+      "[6]My Local Rank      (my_local_rank)\n"
+      "[6]First Proc On Node (first_proc_on_node)\n"
+      "[7]Node Of This PE    (node_of)\n"
+      "[8]Local Rank Of PE   (local_rank_of)\n");
+  start_pe_group_check();
+}
+
+void ParallelInfo::start_pe_group_check() const {
+  Parallel::printf("\n\nStarting Group Check.\n");
+  CProxy_PeGroupReporter::ckNew(CkCallback(
+      CkIndex_ParallelInfo::start_node_group_check(), this->thisProxy));
+}
 
-extern "C" void CkRegisterMainModule(void) {}
+void ParallelInfo::start_node_group_check() const {
+  CProxy_NodeGroupReporter::ckNew(
+      CkCallback(CkIndex_ParallelInfo::end_report(), this->thisProxy));
+}
 
-int main() {
-  printf("%s", info_from_build().c_str());
-  return 0;
+[[noreturn]] void ParallelInfo::end_report() const {
+  Parallel::printf("\nReport finished.\n");
+  Parallel::exit();
 }
+
+// Helper print function so that both node and pe group prints can be changed
+// easily.
+void print_info() {
+  auto compute_digits = [](const auto number) {
+    return number == 0 ? 1 : static_cast<decltype(number)>(
+                                 std::ceil(std::log10(std::abs(number))));
```

Comment:
Incorrect for powers of 10.  Should be `std::ceil(std::log10(std::abs(number) + 1))`.